contracts/stake/src/state.rs

```diff
@@ -0,0 +1,301 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+//
+// Copyright (c) DUSK NETWORK. All rights reserved.
+
+use crate::*;
+
+use alloc::collections::{BTreeMap, BTreeSet};
+use alloc::vec::Vec;
+
+use dusk_bls12_381_sign::PublicKey;
+use dusk_bytes::Serializable;
+use rusk_abi::State;
+
+use stake_contract_types::*;
+use transfer_contract_types::*;
+
+/// Contract keeping track of each public key's stake.
+///
+/// A caller can stake Dusk, and have it attached to a public key. This stake
+/// has a maturation period, after which it is considered valid and the key
+/// eligible to participate in the consensus.
+///
+/// Rewards may be received by a public key regardless of whether they have a
+/// valid stake.
+#[derive(Debug, Default, Clone)]
+pub struct StakeState {
+    stakes: BTreeMap<[u8; PublicKey::SIZE], StakeData>,
+    allowlist: BTreeSet<[u8; PublicKey::SIZE]>,
```

Comment:
The current `master` impl collapse the allowlist inside the stake. This is done in order to prevent double serde

It relies on the fact that any PK inside the stakes is allowed, and the insertion is only performed during the allow tx

Does worth to implement the same pattern here?

Reply:
I understand this optimization pertains to `canonical`? This is no longer (as much) of a problem with this stack, so I would forestall such an optimization to the when we're merging the `piecrust` branch on top of `master`.